apps/emqx_gcp_device/src/emqx_gcp_device.erl

```diff
@@ -0,0 +1,253 @@
+%% Management: import
+-export([import_devices/1]).
+%% Authentication
+-export([get_device_actual_keys/1]).
+%% Internal API
+-export([create_table/0, clear_table/0, format_device/1, config_topic/1]).
+
+-type deviceid() :: binary().
+-type project() :: binary().
+-type location() :: binary().
+-type registry() :: binary().
+-type device_loc() :: {project(), location(), registry()}.
+-type key_type() :: binary().
+-type key() :: binary().
+-type expires_at() :: pos_integer().
+-type key_record() :: {key_type(), key(), expires_at()}.
+-type created_at() :: pos_integer().
+-type extra() :: map().
+
+-record(emqx_gcp_device, {
+    id :: deviceid(),
+    keys :: [key_record()],
+    device_loc :: device_loc(),
+    created_at :: created_at(),
+    extra :: extra()
+}).
+-type emqx_gcp_device() :: #emqx_gcp_device{}.
+
+-type formatted_key() ::
+    #{
+        key_type := key_type(),
+        key := key(),
+        expires_at := expires_at()
+    }.
+-type encoded_config() :: binary().
+-type formatted_device() ::
+    #{
+        deviceid := deviceid(),
+        keys := [formatted_key()],
+        config := encoded_config(),
+        project => project(),
+        location => location(),
+        registry => registry(),
+        created_at => created_at()
+    }.
+-export_type([formatted_device/0, deviceid/0, encoded_config/0]).
+
+-define(TAB, ?MODULE).
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+-spec put_device(formatted_device()) -> ok.
+put_device(FormattedDevice) ->
+    try
+        transaction(fun() -> put_device_no_transaction(FormattedDevice) end)
+    catch
+        _Error:Reason ->
+            ?SLOG(error, #{
+                msg => "Failed to put device",
+                device => FormattedDevice,
+                reason => Reason
+            }),
+            {error, Reason}
+    end.
+
+-spec get_device(deviceid()) -> {ok, formatted_device()} | not_found.
+get_device(DeviceId) ->
+    case ets:lookup(?TAB, DeviceId) of
+        [] ->
+            not_found;
+        [Device] ->
+            {ok, format_device(Device)}
+    end.
+
+-spec remove_device(deviceid()) -> ok.
+remove_device(DeviceId) ->
+    ok = transaction(fun() -> mria:dirty_delete({?TAB, DeviceId}) end),
+    ok = put_config(DeviceId, <<>>).
+
+-spec get_device_actual_keys(deviceid()) -> [key()] | not_found.
+get_device_actual_keys(DeviceId) ->
+    try ets:lookup(?TAB, DeviceId) of
+        [] ->
+            not_found;
+        [Device] ->
+            actual_keys(Device)
+    catch
+        error:badarg ->
+            not_found
+    end.
+
+-spec import_devices([formatted_device()]) ->
+    {NumImported :: non_neg_integer(), NumError :: non_neg_integer()}.
+import_devices(FormattedDevices) when is_list(FormattedDevices) ->
+    transaction(fun() -> lists:foldl(fun import_device/2, {0, 0}, FormattedDevices) end).
```

Comment:
```suggestion
    lists:foldl(fun import_device/2, {0, 0}, FormattedDevices).
```
?

Reply:
Since there's no need for a transaction, maybe it could reduce the number of operations by using `mria:async_dirty` (or `mria:sync_dirty`) as the linked PR does?

https://github.com/emqx/emqx/pull/11389/files#diff-dfb046d7c8660f7ef165880663126a5975583aa1c6e7ff57245d3ee77b1abb63R302

Reply:
Now I see that "`transaction`" has been changed to call one of those functions.  :joy_cat: 

As a minor nit, I'd try to rename it to something else as it's not a transaction, I think.  Maybe `perform_dirty` or similar. :thinking: